### ISSUE TITLE
fire from the crash visible in replay even before the accident

### ISSUE THREAD

**Issue by pedromorgan** [pedromorgan/flightgear-issues-test#1090]
_From [acidzen](https://code.google.com/u/acidzen/) on March 26, 2013 01:45:25_

*What steps will reproduce the problem?* 1. fly, then crash
2. activate replay
3. notice the fire and smoke before the crash moment *What is the expected output? What do you see instead?* fire and smoke should appear after the crash, not before *Any output in the console (black window)?* *What FlightGear version are you using (when using GIT version, please mention date)?* 2.10 *What operating system and graphics card?* linux mint, ati radeon 6800 *Please provide any additional information below or as attachment (Avoid expiring external links, such as to imageshack/pastebin/...).* the same thing seems to happen with rain... if the rain starts during the replay, it will show even if there was no rain during the actual flight

**Attachment:** [fgfs-screen-011.jpg](http://code.google.com/p/flightgear-bugs/issues/detail?id=1091)

_Original issue: http://code.google.com/p/flightgear-bugs/issues/detail?id=1091_

**Comment by pedromorgan** [comment]
_From [gijsrooy](https://code.google.com/u/gijsrooy/) on March 26, 2013 03:26:32_

**Labels:** Type-FeatureRequest

**Comment by pedromorgan** [comment]
_From [wonco...@googlemail.com](https://code.google.com/u/107832458638100707460/) on March 26, 2013 15:31:22_

the replay system now uses the property recorder subsystem to determine which properties to sample, see $FG_ROOT/Docs/README.flightrecorder.

Usually, the configuration will be set up such that only aircraft-specific properties will be sampled/recorded.

Many of the features now found to be incompatible with the flight recorder system, simply predate its existence.

In other words, without explicitly supporting other features, it is to be expected that non-aircraft related events are not properly recorded/replayed, including environmental settings (weather, rain), AI traffic, ATC instructions, wildfire or other "effects" (bombable).

In general, other features like the weather or AI system would need to be modified in order to explicitly support the replay/flight recorder feature.

It's just one of many examples where the FlightGear experience is pretty inconsistent unfortunately.

You may be able to work around some of these issues by creating a custom flight recorder configuration file to sample some of the relevant properties - however, in general, adding replay support to other features, requires coding skills unfortunately.

You will notice that FlightGear exhibits such inconsistencies whenever it needs to synchronize state with other components, not just other subsystems, but also other FlightGear instances, which is the reason why AI traffic/weather are hard to synchronize between multiple sessions currently.

**Comment by pedromorgan** [comment]
_From [wonco...@googlemail.com](https://code.google.com/u/107832458638100707460/) on March 26, 2013 16:29:54_

There are probably more scenarios where inter-instance synchronization simply won't work currently, be it in the form of weather/ai/atc or visual effects for dual control and multi-computer setups in a master/slave fashion.

The current method works such that individual properties need to be individually synchronized using a custom protocol (see fgpanel), which is usually also aircraft-specific, either written in C++ or using the generic protocol system instead.

The flight recorder system could be a viable foundation to improve things here, by using its property sampling wrappers to synchronize complete property tree branches with another instance using the multiplayer XDR handlers via sockets.

For instance, the canvas system is another new system that will otherwise not be "easy" to synchronize among "dual control" clients or multi-machine setups as used during FSWeekend/LinuxTag, even though we will have an increasing number of complex aircraft (airliners) with possibly dozens of canvas-driven MFDs, which may need to support either dual-control or multi-machine configurations with properly synchronized views.

Basically, you would need a way to completely mirror a certain property branch from a master instance onto other slaved fgfs instances, so that things like dual control, advanced weather, AI, or the canvas system can be properly synchronized among different machines.

For the canvas system this could be also pretty relevant, because it being purely property-driven, it would actually be possible to easily replicate a fully working GUI in other instances by sync'ing a property branch, or even have dedicated fgfs instances working as GUI slaves to control the master instance remotely.

**Comment by pedromorgan** [comment]
_From [acidzen](https://code.google.com/u/acidzen/) on March 26, 2013 17:53:23_

Thanks for the explanation. Though my coding barely stretches to some python, it's nice to know what's under the hood and how the things are organized.

**Comment by pedromorgan** [comment]
_From [wonco...@googlemail.com](https://code.google.com/u/107832458638100707460/) on March 27, 2013 03:03:15_

You could still try to solve this particular issue on an individual basis for the aircraft that you are interested in, i.e. by looking at $FG_ROOT/Docs/README.flightrecorder in order to customize the flight recorder configuration such that it also records the crash related properties. 

Assuming that this is an aircraft-specific animation, it should definitely be do-able and not require any coding skills (only XML editing), even though understanding code can be helpful.

**Comment by pedromorgan** [comment]
_From [curtol...@gmail.com](https://code.google.com/u/102377212811568752854/) on March 27, 2013 04:53:04_

For whatever it's worth, fires are not purely property driven.  Properties are used to start or quench a cell, but the propagation is all done in an internal nasal script which maintains the state of all the cells.

**Comment by pedromorgan** [comment]
_From [wonco...@googlemail.com](https://code.google.com/u/107832458638100707460/) on March 27, 2013 06:24:13_

yes, you're referring to the wildfire celluar automaton - but it might be possible to prepare the script such that it resets itself once the replay system is running, so that it can restart - to look at least half-way plausible.

**Comment by pedromorgan** [comment]
_From [curtol...@gmail.com](https://code.google.com/u/102377212811568752854/) on March 27, 2013 06:39:03_

Right, the point was that sending the original bug reporter off to chase property logging can't fully solve the situation.  The cellular automaton nasal script would also need to be rewritten/restructured with a mind towards (a) keeping distributed systems in sync -- even late joiners, and (b) allowing full replays of the fire.

The wildfire simulation was never written with (a) and (b) in mind.  It's hard to call this a bug, because what we are asking for here is some complex interactive behavior that is beyond the scope the original author envisioned.  That said, these are both interesting, useful, possibly important aspects of the simulation, so it might be a good place for one of the developers to invest some effort.

**Comment by pedromorgan** [comment]
_From [wonco...@googlemail.com](https://code.google.com/u/107832458638100707460/) on March 27, 2013 08:15:06_

I wasn't sending anybody off to "chase property logging" at all, I explained right from the beginning how the replay/recorder system works and that there are many features which are simply not supported by it, for all the reasons you mention, too.
It was also explicitly said that workarounds may be possible but that they would usually require coding skills. So I don't think we actually disagree here, or do we ?

And I'm also not sure if anybody called this a bug, it was classified as a feature request earlier and referred to as an "inconsistency". 
And as we know, it's definitely not just the wildfire system that is affected here. 

We have a plethora of other features that no longer work properly once replay comes into play, but also whenever multiple fgfs instances need to be synchronized - be it via multiplayer or master/slave setups.

So this is neither a pure wildfire issue, nor is it specific to the replay/recorder subsystem.

Now, we can argue all day concerning semantics and details, but as can be seen in comment `#3`, I think it would be a little short-sighted to merely emphasize how certain systems were never designed with these features in mind:

Admittedly, many features predate the new flight recorder subsystem - still, the underlying issue is a more general one once we take a step back and look at things from a distance, i.e. what the real problem is and how it is solved in the simulation industry:

The challenge is explicitly managing runtime state in a fashion such that it can be either replicated/synchronized or serialized/reproduced at a later time, without hard-coding the implicit assumption of having just a single main FG instance or a single main FG aircraft, or a continuous simulation time.

We do already have multiple competing approaches here which more or less work to a certain degree, including the generic protocol, multiplayer, the replay/flight recorder system, native protocol and probably a bunch of other options that I don't remember - but they are all sort of "broken" when it comes to more recent features like the canvas, the new weather system, AI/ATC, wildfire etc
Because none of these existing synchronization mechanisms is suitable to synchronize several fgfs instances to ensure that these features work properly in a distributed/replay scenario.

Thus, it would seem obvious that this is an architectural issue that is not specific to certain features, but which would need to be addressed at the design level, by providing an infrastructure to properly synchronize/replicate and serialize state at a later time, and not by fixing individual features one by one.

At the moment, FlightGear provides a bunch of great features that are awesome in isolation, but which often end up crippling the simulator once they are used in combination with a bunch of other options. In fact, you only need to select a certain combination of features, and the simulator won't start up at all.

That's unfortunate, but the truth - and can be seen in various discussions on the devel list, too.

Now, fixing a dozen features like AI/ATC, wildfire, canvas,  weather etc would probably be more work than tackling the underlying design issue, i.e. making the dataflow dependencies really explicit, so that the two problematic main use-cases, i.e. master/slave setups, and data serialization for later replay, become core concepts of the property tree, instead of corner cases, that are only supported on a case-by-case basis.

That way, properties in the property tree could be marked as "master" or "slave" (e.g. via optional attributes) so that a running synchronization mechanism would exactly know if it should compute a property locally, or get a copy from another master instance, or transmit a computed property to slaves. The latter could be even implemented using a simple PUB/SUB protocol implemented on top of the existing telnet/props interface probably.

Another property attribute would determine if/how certain properties need to be serialized, i.e. for replay.

There's some great code and features to be found in FG, especially the original --native protocol stuff that was available already a decade ago when all other simulators still would not support such master/slave use-cases, but somehow the lack of a generic synchronization mechanism is causing more and more breakage, because new features are built on top of a platform that is no longer suitable to support master/slave scenarios or even just replay of full simulator state.
